src/cdk-experimental/popover-edit/table-directives.ts

```diff
@@ -0,0 +1,263 @@
+ */
+@Directive({
+  selector: 'table[editable], cdk-table[editable], mat-table[editable]',
+  providers: [EditEvents],
+})
+export class CdkEditable implements AfterViewInit, OnDestroy {
+  protected readonly destroyed = new ReplaySubject<void>();
+
+  constructor(
+      protected readonly elementRef: ElementRef,
+      protected readonly events: EditEvents,
+      protected readonly ngZone: NgZone) {}
+
+  ngAfterViewInit() {
+    const element = this.elementRef.nativeElement!;
+
+    const toClosest = (selector: string) =>
+        map((event: UIEvent) => closest(event.target, selector));
+
+    this.ngZone.runOutsideAngular(() => {
+      fromEvent<MouseEvent>(element, 'mouseover').pipe(
+          takeUntil(this.destroyed),
+          toClosest(ROW_SELECTOR),
+          ).subscribe(this.events.hovering);
+      fromEvent<MouseEvent>(element, 'mouseleave').pipe(
+          takeUntil(this.destroyed),
+          mapTo(null),
+          ).subscribe(this.events.hovering);
+      fromEvent<MouseEvent>(element, 'mousemove').pipe(
+          takeUntil(this.destroyed),
+          debounceTime(DEFAULT_MOUSE_MOVE_DELAY_MS),
+          toClosest(ROW_SELECTOR),
+          ).subscribe(this.events.mouseMove);
+
+      fromEvent<KeyboardEvent>(element, 'keyup').pipe(
+          takeUntil(this.destroyed),
+          filter(event => event.key === 'Enter'),
+          toClosest(CELL_SELECTOR),
+          ).subscribe(this.events.editing);
+    });
+  }
+
+  ngOnDestroy() {
+    this.destroyed.next();
+    this.destroyed.complete();
+  }
+}
+
+/**
+ * Attaches an ng-template to a cell and shows it when instructed to by the
+ * EditEvents service.
+ * Makes the cell focusable.
+ */
+@Directive({
+  selector: '[cdkPopoverEdit]',
+  host: {
+    'tabIndex': '0',
+    '[attr.aria-haspopup]': 'true',
+  }
+})
+export class CdkPopoverEdit implements AfterViewInit, OnDestroy {
+  /** The edit lens template shown over the cell on edit. */
+  @Input('cdkPopoverEdit') template: TemplateRef<any>|null = null;
+
+  /**
+   * Implicit context to pass along to the template. Can be omitted if the template
+   * is defined within the cell.
+   */
+  @Input('cdkPopoverEditContext') context?: any;
+
+  protected focusTrap?: FocusTrap;
+  protected overlayRef?: OverlayRef;
+  protected readonly destroyed = new ReplaySubject<void>();
+
+  constructor(
+      protected readonly editEvents: EditEvents,
+      protected readonly elementRef: ElementRef,
+      protected readonly focusTrapFactory: FocusTrapFactory,
+      protected readonly ngZone: NgZone,
+      protected readonly overlay: Overlay,
+      protected readonly positionFactory: PopoverEditPositionStrategyFactory,
+      protected readonly viewContainerRef: ViewContainerRef, ) {}
+
+  ngAfterViewInit() {
+    this.editEvents.editingCell(this.elementRef.nativeElement!)
+        .pipe(takeUntil(this.destroyed))
+        .subscribe((open) => {
+          this.ngZone.run(() => {
+            if (open && this.template) {
```

Comment:
Could this chuck of code inside `ngZone.run` be broken into a couple different methods? I think I see some methods like `_createEditOverlay`, `_focusEditOverlay`, and maybe an `_applyFocusTrap` method.